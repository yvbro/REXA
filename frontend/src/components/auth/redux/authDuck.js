import axios from 'axios';
import 'react-toastify/dist/ReactToastify.css';

import { fulfilled, pending, rejected } from '../../../helpers/promise';

const initialState = {
    authenticated: false,
    currentUser: null,
    isAdmin: false,
    loading: false,
};

const LOGIN = '[Auth] LOGIN';
const LOGOUT = '[Auth] LOGOUT';
const FETCH_USER = '[Auth] FETCH_USER';

export default function auth(state = initialState, action) {
    switch (action.type) {
        case pending(LOGIN):
            return {
                ...state,
                authenticated: false,
                currentUser: null,
                isAdmin: false,
                loading: true,
            };
        case fulfilled(LOGIN):
            return {
                ...state,
                authenticated: true,
                currentUser: action.payload.data,
                isAdmin: isAdmin(action.payload.data.roles),
                loading: false,
            };
        case rejected(LOGIN):
            return {
                ...state,
                authenticated: false,
                currentUser: null,
                isAdmin: false,
                loading: false,
            };
        case pending(LOGOUT):
            return {
                ...state,
                authenticated: state.authenticated,
                currentUser: state.currentUser,
                isAdmin: false,
                loading: true,
            };
        case fulfilled(LOGOUT):
            return {
                ...state,
                authenticated: false,
                currentUser: null,
                isAdmin: false,
                loading: false,
            };
        case rejected(LOGOUT):
            return {
                ...state,
                authenticated: state.authenticated,
                currentUser: state.currentUser,
                isAdmin: state.isAdmin,
                loading: false,
            };
        case pending(FETCH_USER):
            return {
                ...state,
                authenticated: state.authenticated,
                currentUser: state.currentUser,
                isAdmin: state.isAdmin,
                loading: true,
            };
        case fulfilled(FETCH_USER):
            return {
                ...state,
                authenticated: true,
                currentUser: action.payload.data,
                isAdmin: isAdmin(action.payload.data.roles),
                loading: false,
            };
        case rejected(FETCH_USER):
            return {
                ...state,
                authenticated: false,
                currentUser: null,
                isAdmin: false,
                loading: false,
            };
        default:
            return state;
    }
}

function isAdmin(authorities) {
    return authorities.includes( 'ADMIN')
}

export const performLogin = (email, password) => (dispatch) =>
    dispatch({
        type: LOGIN,
        payload: axios.post('/auth/login', {
            email: email,
            password: password,
        }),
    });

export const performLogout = () => (dispatch) =>
    dispatch({
        type: LOGOUT,
        payload: axios.post('/auth/logout'),
    });

export const getCurrentUser = () => (dispatch) =>
    dispatch({
        type: FETCH_USER,
        payload: axios.get('/auth/userinfo'),
    });
