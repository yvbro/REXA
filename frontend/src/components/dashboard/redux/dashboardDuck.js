import axios from 'axios';

import { fulfilled, pending, rejected } from '../../../helpers/promise';

const initialState = {
    recentActivities: {
        data: [],
        loading: false,
    },
    preArchives: {
        data: [],
        loading: false,
    },
};

const FETCH_RECENT_ACTIVITIES = '[dashboard] FETCH RECENT ACTIVITIES';
const FETCH_PRE_ARCHIVE = '[dashboard] FETCH PRE ACHIVE';

export default function dashboard(state = initialState, action) {
    switch (action.type) {
        case pending(FETCH_RECENT_ACTIVITIES):
            return {
                ...state,
                recentActivities: {
                    data: [],
                    loading: true,
                },
            };
        case fulfilled(FETCH_RECENT_ACTIVITIES):
            return {
                ...state,
                recentActivities: {
                    data: action.payload.data,
                    loading: false,
                },
            };
        case rejected(FETCH_RECENT_ACTIVITIES):
            return {
                ...state,
                recentActivities: {
                    data: state.recentActivities.data,
                    loading: false,
                },
            };
        case pending(FETCH_PRE_ARCHIVE):
            return {
                ...state,
                preArchives: {
                    data: [],
                    loading: true,
                },
            };
        case fulfilled(FETCH_PRE_ARCHIVE):
            return {
                ...state,
                preArchives: {
                    data: action.payload.data,
                    loading: false,
                },
            };
        case rejected(FETCH_PRE_ARCHIVE):
            return {
                ...state,
                preArchives: {
                    data: state.preArchives.data,
                    loading: false,
                },
            };
        default:
            return state;
    }
}

export const fetchRecentActivities = () => (dispatch) =>
    dispatch({
        type: FETCH_RECENT_ACTIVITIES,
        payload: axios.get('/private/recentActivities'),
    });

export const fetchPreAchives = () => (dispatch) =>
    dispatch({
        type: FETCH_PRE_ARCHIVE,
        payload: axios.get('/private/preArchives'),
    });
