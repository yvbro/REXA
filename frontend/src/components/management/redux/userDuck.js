import axios from 'axios';

import { toast } from 'react-toastify';
import _get from 'lodash/get';

import { fulfilled, pending, rejected } from '../../../helpers/promise';

const DEFAULT_ROLES = ['USER'];
const initialState = {
    data: [],
    loading: false,
};

const setEnabledForUser = (data, userEmail, enabled) => {
    for (const i in data) {
        if (data[i].email === userEmail) {
            data[i].enabled = enabled;
            break;
        }
    }
    return data;
};

const addNewUser = (data, userEmail) => {
    data.push({ email: userEmail, enabled: false, roles: DEFAULT_ROLES });
    return data;
};

const FETCH_USERS = '[User] FETCH LIST OF USERS';
const SWITCH_ENABLED_USER = '[User] EDIT ENABLED USER';
const ADD_USER = '[User] ADD USER';

export default function project(state = initialState, action) {
    switch (action.type) {
        case pending(FETCH_USERS):
            return {
                ...state,
                data: [],
                loading: true,
            };
        case fulfilled(FETCH_USERS):
            return {
                ...state,
                data: action.payload.data,
                loading: false,
            };
        case rejected(FETCH_USERS):
            return {
                ...state,
                data: state.data,
                loading: false,
            };
        case pending(SWITCH_ENABLED_USER):
            return {
                ...state,
                data: setEnabledForUser(
                    state.data,
                    action.payload.userEmail,
                    action.payload.enabled
                ),
                loading: false,
            };
        case rejected(SWITCH_ENABLED_USER):
            return {
                ...state,
                data: setEnabledForUser(
                    state.data,
                    action.payload.userEmail,
                    !action.payload.enabled
                ),
                loading: false,
            };
        case ADD_USER:
            return {
                ...state,
                data: addNewUser(state.data, action.payload),
                loading: false,
            };
        default:
            return state;
    }
}

export const fetchUsers = () => (dispatch) =>
    dispatch({
        type: FETCH_USERS,
        payload: axios.get('/private/management/users'),
    });

export const switchEnabledUser = (userEmail, enabled) => (dispatch) => {
    const param = { userEmail: userEmail, enabled: enabled };

    dispatch({
        type: pending(SWITCH_ENABLED_USER),
        payload: param,
    });

    return axios
        .post(`/private/management/users/switch`, param)
        .then(() => toast.info('Changes saved!'))
        .catch(() => {
            dispatch({
                type: rejected(SWITCH_ENABLED_USER),
                payload: param,
            });
            toast.error('Your changes could not be saved.');
        });
};

export const addUser = (email, password) => (dispatch) => {
    const param = { email: email, password: password };

    return axios
        .post(`/private/management/users/add`, param)
        .then(() => {
            dispatch({
                type: ADD_USER,
                payload: email,
            });
            toast.info('User added!');
        })
        .catch((error) => {
            let errorMessage = _get(error, 'response.data.message', null);
            toast.error(errorMessage);
        });
};
