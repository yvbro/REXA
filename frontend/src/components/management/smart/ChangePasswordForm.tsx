import React, { useState } from 'react';

import {
    Card,
    CardHeader,
    CardContent,
    CardActions,
    Button,
    makeStyles,
    Avatar,
} from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { toast } from 'react-toastify';
import PasswordRules from '../../common/password/PasswordRules';
import PasswordField from '../../common/password/PasswordField';

import {
    isPasswordTooShort,
    ERROR_PASSWORD_LENGTH,
    passwordDoesNotContainACapitalLetter,
    ERROR_PASSWORD_CAPITAL_LETTER,
    passwordDoesNotContainANumber,
    ERROR_PASSWORD_NUMBER,
    ERROR_PASSWORD_NOT_MATCH,
} from '../../../helpers/constants';

import themes from '../../common/theme/theme.scss';
import useUsersManagementService from '../../../services/useUsersManagementService';

const useStyles = makeStyles((theme) => ({
    root: {
        margin: 'auto',
        paddingTop: 20,
        width: 500,
        height: 400,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        flexDirection: 'column',
        borderRadius: themes.borderRadius,
    },
    header: {
        width: 430,
    },
    text: {
        paddingBottom: 10,
        width: 400,
    },
    cardContent: {
        display: 'flex',
        flexDirection: 'column',
    },
    saveButton: {
        width: 120,
        backgroundColor: themes.primaryButtonColor,
        color: 'white',
    },
    cancelButton: {
        width: 120,
        backgroundColor: themes.secondaryButtonColor,
        color: 'white',
    },
    avatar: {
        width: theme.spacing(8),
        height: theme.spacing(8),
        backgroundColor: themes.avatarColor,
    },
    iconDef: {
        fontSize: 50,
    },
}));

interface ChangePasswordFormProps {
    userEmail: string;
    closeAction: () => void;
}

function ChangePasswordForm({ userEmail, closeAction }: ChangePasswordFormProps) {
    const classes = useStyles();

    const { updatePassword } = useUsersManagementService();

    const [newPassword, setNewPassword] = useState('');
    const [confirmationPassword, setConfirmationPassword] = useState('');
    const [errorPassword, setErrorPassword] = useState('');
    const [errorConfirmationPassword, setErrorConfirmationPassword] = useState('');

    const handleSubmit = async (event: React.FormEvent) => {
        event.preventDefault();

        if (confirmationPassword !== newPassword) {
            setErrorConfirmationPassword(ERROR_PASSWORD_NOT_MATCH);
        } else {
            updatePassword({
                email: userEmail,
                newPassword,
                confirmationPassword,
            });
            closeAction();
        }
    };

    const onChangePassword = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (isPasswordTooShort(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_LENGTH);
        } else if (passwordDoesNotContainACapitalLetter(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_CAPITAL_LETTER);
        } else if (passwordDoesNotContainANumber(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_NUMBER);
        } else {
            setErrorPassword('');
        }

        setNewPassword(event.target.value);
    };

    const onChangeConfirmedPassword = (
        event: React.ChangeEvent<HTMLInputElement>
    ) => {
        if (event.target.value !== newPassword) {
            setErrorConfirmationPassword(
                'Password and confirmation does not match.'
            );
        } else {
            setErrorConfirmationPassword('');
        }
        setConfirmationPassword(event.target.value);
    };

    return (
        <Card className={classes.root}>
            <CardHeader
                avatar={
                    <Avatar aria-label="recipe" className={classes.avatar}>
                        <EditIcon className={classes.iconDef} />
                    </Avatar>
                }
                title={`Edit password for ${userEmail}`}
                titleTypographyProps={{ variant: 'h6' }}
                subheader={<PasswordRules />}
                className={classes.header}
            />
            <CardContent className={classes.cardContent}>
                <PasswordField
                    value={newPassword}
                    label="New password"
                    error={errorPassword}
                    onChange={onChangePassword}
                    testId="newPassword"
                    className={classes.text}
                />
                <PasswordField
                    value={confirmationPassword}
                    label="Confirm new password"
                    error={errorConfirmationPassword}
                    onChange={onChangeConfirmedPassword}
                    testId="confirmationPassword"
                    className={classes.text}
                />
            </CardContent>
            <CardActions>
                <Button
                    variant="contained"
                    onClick={handleSubmit}
                    disabled={
                        !newPassword ||
                        !confirmationPassword ||
                        !!errorPassword ||
                        !!errorConfirmationPassword
                    }
                    className={classes.saveButton}
                >
                    Save
                </Button>
                <Button
                    variant="contained"
                    onClick={closeAction}
                    className={classes.cancelButton}
                >
                    Cancel
                </Button>
            </CardActions>
        </Card>
    );
}

export default ChangePasswordForm;
