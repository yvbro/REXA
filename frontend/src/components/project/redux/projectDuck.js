import axios from 'axios';
import 'react-toastify/dist/ReactToastify.css';

import { fulfilled, pending, rejected } from '../../../helpers/promise';

const initialState = {
    projectsList: {
        data: [],
        loading: false,
    },
    selectedProject: {
        data: undefined,
        loading: false,
    },
};

const FETCH_PROJECTS = '[Project] FETCH LIST OF PROJECTS';
const FETCH_PROJECT = '[Project] FETCH ONE PROJECT';
const RESET_DATA_PROJECTS = '[dashboard] RESET DATA PROJECTS';

export default function project(state = initialState, action) {
    switch (action.type) {
        case pending(FETCH_PROJECTS):
            return {
                ...state,
                projectsList: {
                    data: [],
                    loading: true,
                },
            };
        case fulfilled(FETCH_PROJECTS):
            return {
                ...state,
                projectsList: {
                    data: action.payload.data.sort((a, b) =>
                        a.name.localeCompare(b.name)
                    ),
                    loading: false,
                },
            };
        case rejected(FETCH_PROJECTS):
            return {
                ...state,
                projectsList: {
                    data: state.projectsList.data,
                    loading: false,
                },
            };
        case pending(FETCH_PROJECT):
            return {
                ...state,
                selectedProject: {
                    data: [],
                    loading: true,
                },
            };
        case fulfilled(FETCH_PROJECT):
            return {
                ...state,
                selectedProject: {
                    data: action.payload.data,
                    loading: false,
                },
            };
        case rejected(FETCH_PROJECT):
            return {
                ...state,
                selectedProject: {
                    data: state.selectedProject.data,
                    loading: false,
                },
            };
        case RESET_DATA_PROJECTS:
            return {
                projectsList: {
                    data: [],
                    loading: false,
                },
                selectedProject: {
                    data: undefined,
                    loading: false,
                },
            };
        default:
            return state;
    }
}

export const fetchProjects = () => (dispatch) =>
    dispatch({
        type: FETCH_PROJECTS,
        payload: axios.get('/private/projects'),
    });

export const fetchProject = (projectId) => (dispatch) =>
    dispatch({
        type: FETCH_PROJECT,
        payload: axios.get(`/private/projects/${projectId}`),
    });

export const resetDataProjects = () => (dispatch) =>
    dispatch({
        type: RESET_DATA_PROJECTS,
    });
