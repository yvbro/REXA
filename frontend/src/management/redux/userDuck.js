import axios from 'axios';

import { fulfilled, pending, rejected } from '../../helpers/promise';

const initialState = {
    users: {
        data: [],
        loading: false,
    },
};

const FETCH_USERS = '[Auth] FETCH LIST OF USERS';

export default function project(state = initialState, action) {
    switch (action.type) {
        case pending(FETCH_USERS):
            return {
                ...state,
                users: {
                    data: [],
                    loading: true,
                },
            };
        case fulfilled(FETCH_USERS):
            return {
                ...state,
                users: {
                    data: action.payload.data,
                    loading: false,
                },
            };
        case rejected(FETCH_USERS):
            return {
                ...state,
                users: {
                    data: state.user.data,
                    loading: false,
                },
            };
        default:
            return state;
    }
}

export const fetchUsers = () => (dispatch) =>
    dispatch({
        type: FETCH_USERS,
        payload: axios.get('/private/management/users'),
    });
