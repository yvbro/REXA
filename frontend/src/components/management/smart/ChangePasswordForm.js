import React, { useState } from 'react';
import { PropTypes } from 'prop-types';

import {
    Card,
    CardHeader,
    CardContent,
    CardActions,
    Button,
    makeStyles,
    Avatar,
} from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import PasswordRules from '../../common/PasswordRules';
import PasswordField from '../../common/PasswordField';

import { updatePassword } from '../redux/userDuck';
import {
    isPasswordTooShort,
    ERROR_PASSWORD_LENGTH,
    passwordDoesNotContainACapitalLetter,
    ERROR_PASSWORD_CAPITAL_LETTER,
    passwordDoesNotContainANumber,
    ERROR_PASSWORD_NUMBER,
    ERROR_PASSWORD_NOT_MATCH
} from '../../../helpers/constants';

const useStyles = makeStyles((theme) => ({
    root: {
        margin: 'auto',
        paddingTop: 20,
        width: 500,
        height: 400,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        flexDirection: 'column',
        borderRadius: '16px',
    },
    text: {
        paddingBottom: 10,
        width: 400,
    },
    cardContent: {
        display: 'flex',
        flexDirection: 'column',
    },
    button: {
        width: 100,
    },
    avatar: {
        width: theme.spacing(8),
        height: theme.spacing(8),
    },
    iconDef: {
        fontSize: 50,
    },
}));

const ChangePasswordForm = (props) => {
    const classes = useStyles();

    const [newPassword, setNewPassword] = useState('');
    const [confirmationPassword, setConfirmationPassword] = useState('');
    const [errorPassword, setErrorPassword] = useState('');
    const [errorConfirmationPassword, setErrorConfirmationPassword] = useState('');

    const handleSubmit = (event) => {
        event.preventDefault();

        if (confirmationPassword !== newPassword) {
            setErrorConfirmationPassword(ERROR_PASSWORD_NOT_MATCH);
        } else {
            updatePassword(props.userEmail, newPassword, confirmationPassword).then(() => props.closeAction());
        }
    };

    const onChangePassword = (event) => {
        if (isPasswordTooShort(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_LENGTH);
        } else if (passwordDoesNotContainACapitalLetter(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_CAPITAL_LETTER);
        } else if (passwordDoesNotContainANumber(event.target.value)) {
            setErrorPassword(ERROR_PASSWORD_NUMBER);
        } else {
            setErrorPassword('');
        }

        setNewPassword(event.target.value);
    };

    const onChangeConfirmedPassword = (event) => {
        if (event.target.value !== newPassword) {
            setErrorConfirmationPassword('Password and confirmation does not match.');
        } else {
            setErrorConfirmationPassword('');
        }
        setConfirmationPassword(event.target.value);
    };

    return (
        <Card className={classes.root}>
            <CardHeader
                avatar={
                    <Avatar aria-label="recipe" className={classes.avatar}>
                        <EditIcon className={classes.iconDef} />
                    </Avatar>
                }
                title={`Edit password for ${props.userEmail}`}
                titleTypographyProps={{ variant: 'h6', color: 'primary' }}
                subheader={<PasswordRules />}
            />
            <CardContent className={classes.cardContent}>
                <PasswordField 
                    value={newPassword}
                    label="New password"
                    error={errorPassword}
                    onChange={onChangePassword}
                    testId={'newPassword'}
                    className={classes.text}
                />
                <PasswordField 
                    value={confirmationPassword}
                    label="Confirm new password"
                    error={errorConfirmationPassword}
                    onChange={onChangeConfirmedPassword}
                    testId={'confirmationPassword'}
                    className={classes.text}
                />
            </CardContent>
            <CardActions>
                <Button
                    size="small"
                    onClick={handleSubmit}
                    color="primary"
                    disabled={
                        !newPassword || !confirmationPassword ||
                        !!errorPassword || !!errorConfirmationPassword}
                    className={classes.button}
                >
                    Save
                </Button>
                <Button
                    size="small"
                    onClick={props.closeAction}
                    color="secondary"
                    className={classes.button}
                >
                    Cancel
                </Button>
            </CardActions>
        </Card>
    );
};

ChangePasswordForm.propTypes = {
    userEmail: PropTypes.string,
    closeAction: PropTypes.func.isRequired,
};

export default ChangePasswordForm;
