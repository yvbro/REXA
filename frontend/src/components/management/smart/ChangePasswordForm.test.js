import React from 'react';
import { cleanup, fireEvent, renderWithStore } from '../../../helpers/test/test-utils';
import '@testing-library/jest-dom/extend-expect';

import ChangePasswordForm from './ChangePasswordForm';

import * as userDuck from '../redux/userDuck';

// With jest.mock our API method does nothing
// we don't want to hit the server in our tests
jest.mock('../redux/userDuck');

const CLOSE_ACTION = jest.fn();

const VALID_PASSWORD = 'Qwer1234';
const SHORT_PASSWORD = 'wrong';
const NO_NUMBER_PASSWORD = 'No number password';
const NO_CAPITAL_LETTER_PASSWORD = 'no capital letter password';
const USER_EMAIL = 'test@gmail.com';

describe('The ChangePasswordForm component', () => {
    
    afterEach(cleanup);

    it('should take a snapshot', () => {
        const { asFragment } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        expect(asFragment(<ChangePasswordForm closeAction={CLOSE_ACTION}/>)).toMatchSnapshot();
    });

    it('should contain two buttons with save and cancel', () => {
        const { getAllByRole, getByRole } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        expect(getAllByRole('button')).toHaveLength(2);
        expect(getByRole('button', {name: 'Save'})).toBeInTheDocument();
        expect(getByRole('button', {name: 'Cancel'})).toBeInTheDocument();
    });

    it('should display the password rules', () => {
        const { getByTestId } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        expect(getByTestId('passwordRules')).toHaveTextContent('The password must be 8 characters long. It should contain a capital letter and a number.');
    });

    it('should have button Save disabled if no value', () => {
        const { getByRole } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        expect(getByRole('button', {name: 'Save'})).toBeDisabled();
    });

    it('should call closeAction on Cancel button', () => {
        const { getByRole } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        const cancel = getByRole('button', {name: 'Cancel'});
        fireEvent.click(cancel);

        expect(CLOSE_ACTION).toHaveBeenCalledTimes(1);
    });

    it('should call updatePassword on Save if no errors and close modal', () => {
        const { getByRole, getByTestId } = renderWithStore(<ChangePasswordForm userEmail={USER_EMAIL} closeAction={CLOSE_ACTION}/>, {});

        userDuck.updatePassword = jest.fn(() => Promise.resolve());
        
        const passwordInput = getByTestId('newPassword');
        const confirmationpasswordInput = getByTestId('confirmationPassword');

        fireEvent.change(passwordInput, { target: { value: VALID_PASSWORD } });
        fireEvent.change(confirmationpasswordInput, { target: { value: VALID_PASSWORD } });

        const save = getByRole('button', {name: 'Save'});
        fireEvent.click(save);

        expect(userDuck.updatePassword).toHaveBeenCalledTimes(1);
        expect(userDuck.updatePassword).toHaveBeenCalledWith(USER_EMAIL, VALID_PASSWORD, VALID_PASSWORD);
        expect(CLOSE_ACTION).toHaveBeenCalledTimes(1);
    });

    it('should display error if password to short', () => {
        const { getByTestId, getByText } = renderWithStore(<ChangePasswordForm userEmail={USER_EMAIL} closeAction={CLOSE_ACTION}/>, {});
        
        const passwordInput = getByTestId('newPassword');

        fireEvent.change(passwordInput, { target: { value: SHORT_PASSWORD } });

        expect(getByText('Password must be at least 8 characters long.')).toBeInTheDocument();
    });

    it('should display error if no number in password', () => {
        const { getByTestId, getByText } = renderWithStore(<ChangePasswordForm userEmail={USER_EMAIL} closeAction={CLOSE_ACTION}/>, {});
        
        const passwordInput = getByTestId('newPassword');

        fireEvent.change(passwordInput, { target: { value: NO_NUMBER_PASSWORD } });

        expect(getByText('Password must contain a number.')).toBeInTheDocument();
    });

    it('should display error if no capital letter in password', () => {
        const { getByTestId, getByText } = renderWithStore(<ChangePasswordForm userEmail={USER_EMAIL} closeAction={CLOSE_ACTION}/>, {});
        
        const passwordInput = getByTestId('newPassword');

        fireEvent.change(passwordInput, { target: { value: NO_CAPITAL_LETTER_PASSWORD } });

        expect(getByText('Password must contain a capital letter.')).toBeInTheDocument();
    });

    it('should display error if no match between password and confirmation', () => {
        const { getByTestId, getByText } = renderWithStore(<ChangePasswordForm userEmail={USER_EMAIL} closeAction={CLOSE_ACTION}/>, {});
        
        const passwordInput = getByTestId('newPassword');
        const confirmationpasswordInput = getByTestId('confirmationPassword');

        fireEvent.change(passwordInput, { target: { value: VALID_PASSWORD } });
        fireEvent.change(confirmationpasswordInput, { target: { value: SHORT_PASSWORD } });

        expect(getByText('Password and confirmation does not match.')).toBeInTheDocument();
    });

    it('should have button Save disabled if missing confirmation password', () => {
        const { getByRole, getByTestId } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        const passwordInput = getByTestId('newPassword');

        fireEvent.change(passwordInput, { target: { value: VALID_PASSWORD } });

        expect(getByRole('button', {name: 'Save'})).toBeDisabled();
    });

    it('should have button Save disabled if missing new password', () => {
        const { getByRole, getByTestId } = renderWithStore(<ChangePasswordForm closeAction={CLOSE_ACTION}/>, {});

        const confirmationpasswordInput = getByTestId('confirmationPassword');

        fireEvent.change(confirmationpasswordInput, { target: { value: VALID_PASSWORD } });

        expect(getByRole('button', {name: 'Save'})).toBeDisabled();
    });
});