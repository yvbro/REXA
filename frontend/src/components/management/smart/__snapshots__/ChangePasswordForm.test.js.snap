// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ChangePasswordForm component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          aria-label="recipe"
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-5 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-iconDef-6"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-displayBlock"
        >
          Edit password for undefined
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          <span
            class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom MuiTypography-displayBlock"
            data-testid="passwordRules"
          >
            The password must be 8 characters long. It should contain a capital letter and a number.
          </span>
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root makeStyles-cardContent-3"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-text-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          New password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
            data-testid="newPassword"
            form="[object Object]"
            required=""
            type="password"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              aria-label="toggle password visibility"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9"
            >
              <span>
                New password *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-text-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Confirm new password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
            data-testid="confirmationPassword"
            form="[object Object]"
            required=""
            type="password"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              aria-label="toggle password visibility"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9"
            >
              <span>
                Confirm new password *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-4 MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Save
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Cancel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
